src/bootstrap/dist.rs

```diff
@@ -1344,6 +1357,17 @@ impl Step for Extended {
         builder.info(&format!("Dist extended stage{} ({})", compiler.stage, target));
 
         let mut tarballs = Vec::new();
+        let mut built_tools = HashSet::new();
+        macro_rules! add_tool {
+            ($name:expr => $step:expr) => {
+                if should_build_extended_tool(builder, $name) {
+                    if let Some(tarball) = builder.ensure($step) {
```

Comment:
I think it's a larger refactor, but maybe not actually that hard to do -- can we instead of running builder.ensure here add a `builder.ensure_if_default` or something like it, which runs code similar to `builder.default_doc` (but constrained to just that particular step)?

That way we don't have to repeat the should_build_extended_tool here, and that seems overall nicer. It would also make `--exclude rustfmt` just work, I believe, whereas this currently wouldn't support that (which seems not nice).

Reply:
Pushed the refactoring!